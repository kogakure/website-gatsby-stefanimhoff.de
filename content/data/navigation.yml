- navigation:
    - text: "About"
      url: "/about/"
    - text: "Projects"
      url: "/projects/"
