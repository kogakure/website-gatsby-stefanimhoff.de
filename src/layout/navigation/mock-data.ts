export const navigation = [
  { text: 'About', url: '/about/' },
  { text: 'Projects', url: '/projects/' },
];
