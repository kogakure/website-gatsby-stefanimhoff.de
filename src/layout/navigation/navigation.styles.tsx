import styled from 'styled-components';
import { Link } from 'gatsby';

import { ArrowLeft } from '../../components/icons';

const Navigation = styled.nav`
  align-items: center;
  display: flex;
`;

const BackLink = styled.div`
  align-items: center;
  display: flex;
`;

const List = styled.ul`
  align-items: center;
  display: flex;
  list-style: none;
  margin: 0;
`;

const Item = styled.li`
  margin-right: 20px;
`;

const ItemLink = styled(Link)`
  color: var(--colorText);
  text-decoration: none;
  text-decoration-color: var(--colorLink);

  &:hover,
  &:focus {
    text-decoration: underline;
    text-decoration-color: var(--colorLinkActive) !important;
    text-decoration-thickness: 0.1em;
    text-underline-offset: 0.1em;

    @supports (not (text-decoration-color: red)) {
      color: var(--colorLink);
    }
  }
`;

const Arrow = styled(ArrowLeft)`
  fill: var(--colorText);
  font-size: 1.2rem;
`;

const ArrowLink = styled(Link)`
  align-items: center;
  display: flex;
`;

export const Styled = {
  Arrow,
  ArrowLink,
  BackLink,
  List,
  Item,
  ItemLink,
  Navigation,
};
