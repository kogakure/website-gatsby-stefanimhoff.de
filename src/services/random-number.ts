export const getRandomNumber = (max: number) =>
  String(Math.floor(Math.random() * Math.floor(max)));
