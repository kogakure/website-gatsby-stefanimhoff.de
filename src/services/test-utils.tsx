import React from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';

import { EmojifyProvider, LocalizedDateProvider } from '../contexts';

const customRender = (
  node: React.ReactElement,
  options?: RenderOptions
): RenderResult => {
  const rendered = render(
    <EmojifyProvider>
      <LocalizedDateProvider>{node}</LocalizedDateProvider>
    </EmojifyProvider>,
    options
  );

  return {
    ...rendered,
    rerender: (ui: React.ReactElement, rerenderOptions?: RenderOptions) =>
      customRender(ui, { container: rendered.container, ...options }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
