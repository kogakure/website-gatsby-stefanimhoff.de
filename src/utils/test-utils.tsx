import React from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { EmojifyProvider, LocalizedDateProvider } from '../contexts';
import { ThemeType } from '../typings/Theme';
import { lightTheme } from '../themes';

interface CustomTheme {
  theme: ThemeType;
}

type RenderOptionsWithTheme = RenderOptions & CustomTheme;

const customRender = (
  node: React.ReactElement,
  options: RenderOptionsWithTheme = { theme: lightTheme }
): RenderResult => {
  const { theme } = options;
  const rendered = render(
    <ThemeProvider theme={theme}>
      <EmojifyProvider>
        <LocalizedDateProvider>{node}</LocalizedDateProvider>
      </EmojifyProvider>
    </ThemeProvider>,
    options
  );

  return {
    ...rendered,
    rerender: (
      ui: React.ReactElement,
      rerenderOptions?: RenderOptionsWithTheme
    ) => customRender(ui, { container: rendered.container, ...options }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
