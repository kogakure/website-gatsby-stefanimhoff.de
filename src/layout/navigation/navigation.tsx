import * as React from 'react';
import { useStaticQuery, graphql } from 'gatsby';

import { Clickable } from '../../components/clickable';
import { NavigationQueryData } from '../../typings/navigation-query-data';

import { Styled } from './navigation.styles';

export const Navigation: React.FC = () => {
  const data: NavigationQueryData = useStaticQuery(graphql`
    query NavigationQuery {
      navigationYaml {
        navigation {
          text
          url
        }
      }
    }
  `);

  const { navigation } = data.navigationYaml;

  return (
    <Styled.Navigation className="nav">
      <Styled.BackLink>
        <Styled.ArrowLink to="/" activeStyle={{ visibility: 'hidden' }}>
          <Clickable text="Back to Homepage">
            <Styled.Arrow />
          </Clickable>
        </Styled.ArrowLink>
      </Styled.BackLink>
      <Styled.List>
        {navigation.map((item) => (
          <Styled.Item key={item.text}>
            <Styled.ItemLink
              to={item.url}
              activeStyle={{
                textDecoration: 'underline',
                textDecorationThickness: '0.1em',
                textUnderlineOffset: '0.1em',
                textDecorationColor: 'var(--colorLink)',
              }}
            >
              {item.text}
            </Styled.ItemLink>
          </Styled.Item>
        ))}
      </Styled.List>
    </Styled.Navigation>
  );
};
