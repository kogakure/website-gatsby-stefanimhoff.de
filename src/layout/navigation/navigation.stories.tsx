import React from 'react';

import { navigation } from './mock-data';

import { Navigation } from '.';

export default {
  component: Navigation,
  title: 'Layout/Navigation',
};

export const Default = () => <Navigation navigation={navigation} />;
