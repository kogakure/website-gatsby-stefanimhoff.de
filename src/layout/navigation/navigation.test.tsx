import React from 'react';
import { axe } from 'jest-axe';
import * as Gatsby from 'gatsby';

import { render } from '../../services/test-utils';

import { Navigation } from './navigation';

const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');

useStaticQuery.mockImplementation(() => ({
  navigationYaml: {
    navigation: [
      { text: 'About', url: '/about/' },
      { text: 'Projects', url: '/projects/' },
    ],
  },
}));

describe('Navigation', () => {
  test('renders correctly', () => {
    const { container } = render(<Navigation />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('is accessible', async () => {
    const { container } = render(<Navigation />);
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
