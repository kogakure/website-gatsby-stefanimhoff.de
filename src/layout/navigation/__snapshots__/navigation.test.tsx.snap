// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders correctly 1`] = `
.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: 0;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 40px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 40px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
  margin: 0;
}

.c6 {
  margin-right: 20px;
}

.c7 {
  color: var(--colorText);
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-text-decoration-color: var(--colorLink);
  text-decoration-color: var(--colorLink);
}

.c7:hover,
.c7:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-text-decoration-color: var(--colorLinkActive) !important;
  text-decoration-color: var(--colorLinkActive) !important;
  -webkit-text-decoration-thickness: 0.1em;
  text-decoration-thickness: 0.1em;
  text-underline-offset: 0.1em;
}

.c4 {
  fill: var(--colorText);
  font-size: 1.2rem;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@supports (not (text-decoration-color:red)) {
  .c7:hover,
  .c7:focus {
    color: var(--colorLink);
  }
}

<nav
  class="c0 nav"
>
  <div
    class="c1"
  >
    <a
      class="c2"
      href="/"
    >
      <button
        aria-label="Back to Homepage"
        class="c3"
        type="button"
      >
        <svg
          class="c4"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M7.828 11H20v2H7.828l5.364 5.364-1.414 1.414L4 12l7.778-7.778 1.414 1.414z"
          />
        </svg>
      </button>
    </a>
  </div>
  <ul
    class="c5"
  >
    <li
      class="c6"
    >
      <a
        class="c7"
        href="/about/"
      >
        About
      </a>
    </li>
    <li
      class="c6"
    >
      <a
        class="c7"
        href="/projects/"
      >
        Projects
      </a>
    </li>
  </ul>
</nav>
`;
